Uncaught SyntaxError: Identifier 'i2' has already been declared

When bundling the following code to be used in the browser, I get the error *Uncaught SyntaxError: Identifier 'i2' has already been declared* when loaded in the browser.

### index.html
```html
<script type="module" src="bundle.js"></script>
```

### modules.js
```javascript
import rough from 'https://unpkg.com/roughjs/bundled/rough.esm.js?module';
import { LitElement } from 'https://unpkg.com/lit-element/lit-element.js?module';
export { rough, LitElement }
```

I'm using the following command to bundle the code:
```bash
deno bundle modules.js bundle.js
```

I was using the latest version 1.16.1 to run this, but because I remember this used to work with no errors, I compiled a list of any errors for each version in the table below.

I found that there were no errors from version 1.7.0 until 1.15.0.

The error also goes away if I remove the import of the *rough* module from _modules.js_ which I think also includes modules that are duplicated in the *lit-element* module.

This error looks like it might be related to https://github.com/denoland/deno/issues/8959 which was closed after a fix earlier this year, but both output *Uncaught SyntaxError: Identifier 'i2' has already been declared*

### Windows 10 Linux Subsystem
Build: 22000
Branch: co_release
Release: Ubuntu 20.04.3 LTS,
Kernel: Linux 5.10.16.3-microsoft-standard-WSL2

<!--StartFragment--><google-sheets-html-origin><!--td {border: 1px solid #ccc;}br {mso-data-placement:same-cell;}-->
Version | Error
-- | --
1.16.1 | Uncaught SyntaxError: Identifier 'i2' has already been declared
1.16.0 | Uncaught SyntaxError: Identifier 'i2' has already been declared
1.15.3 | Uncaught SyntaxError: Identifier 'i2' has already been declared
1.15.2 | Uncaught SyntaxError: Identifier 'i2' has already been declared
1.15.1 | Uncaught SyntaxError: Identifier 'i2' has already been declared
1.15.0 |  
1.14.3 |  
1.14.2 |  
1.14.1 |  
1.14.0 |  
1.13.2 |  
1.13.1 |  
1.13.0 |  
1.12.2 |  
1.12.1 |  
1.12.0 |  
1.11.5 |  
1.11.4 |  
1.11.3 |  
1.11.2 |  
1.11.1 |  
1.11.0 |  
1.10.3 |  
1.10.2 |  
1.10.1 |  
1.10.0 | This version can't be found
1.9.2 |  
1.9.1 |  
1.9.0 |  
1.8.3 |  
1.8.2 |  
1.8.1 |  
1.8.0 |  
1.7.5 |  
1.7.4 |  
1.7.3 | This version can't be found
1.7.2 |  
1.7.1 |  
1.7.0 |  
1.6.3 | Uncaught ReferenceError: ReactiveElement is not defined at bundle.js:2075
1.6.2 | Uncaught ReferenceError: ReactiveElement is not defined at bundle.js:2182
1.6.1 | Uncaught ReferenceError: ReactiveElement is not defined at bundle.js:2182
1.6.0 | Uncaught ReferenceError: ReactiveElement is not defined at bundle.js:2182
1.5.4 | Uncaught ReferenceError: ReactiveElement is not defined at bundle.js:1015
1.5.3 | Uncaught ReferenceError: noChange is not defined at bundle.js:1974
1.5.2 | Uncaught ReferenceError: noChange is not defined at bundle.js:1974
1.5.1 | Uncaught ReferenceError: noChange is not defined at bundle.js:1974
1.5.0 | bundle.js:2233 Uncaught SyntaxError: Identifier 'e6' has already been declared
1.4.6 | Can't Compile because .js?module files are considered typscript
1.4.5 | Can't Compile because .js?module files are considered typscript
1.4.4 | Can't Compile because .js?module files are considered typscript
1.4.3 | Can't Compile because .js?module files are considered typscript
1.4.2 | Can't Compile because .js?module files are considered typscript
1.4.1 | Can't Compile because .js?module files are considered typscript
1.4.0 | Can't Compile because .js?module files are considered typscript
1.3.3 | Can't Compile because .js?module files are considered typscript
1.3.2 | Can't Compile because .js?module files are considered typscript
1.3.1 | Can't Compile because .js?module files are considered typscript
1.3.0 | Can't Compile because .js?module files are considered typscript
1.2.3 | Can't Compile because .js?module files are considered typscript
1.2.2 | Can't Compile because .js?module files are considered typscript
1.2.1 | Can't Compile because .js?module files are considered typscript
1.2.0 | Can't Compile because .js?module files are considered typscript
1.1.3 | Can't Compile because .js?module files are considered typscript
1.1.2 | Can't Compile because .js?module files are considered typscript
1.1.1 | Can't Compile because .js?module files are considered typscript
1.1.0 | Can't Compile because .js?module files are considered typscript
1.0.5 | Can't Compile because .js?module files are considered typscript
1.0.4 | Can't Compile because .js?module files are considered typscript
1.0.3 | Can't Compile because .js?module files are considered typscript
1.0.2 | Can't Compile because .js?module files are considered typscript
1.0.1 | error: Import 'https://unpkg.com/@lit/css-tag.js?module' failed: 404 Not Found
1.0.0 | error: Uncaught AssertionError: Assertion failed.
<!--EndFragment-->

--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -39,11 +39,11 @@ winapi = "=0.3.9"
 winres = "=0.1.11"
 
 [dependencies]
-deno_ast = { version = "=0.7.0", features = ["bundler", "codegen", "dep_graph", "module_specifier", "proposal", "react", "sourcemap", "transforms", "typescript", "view", "visit"] }
+deno_ast = { version = "0.7.0", features = ["bundler", "codegen", "dep_graph", "module_specifier", "proposal", "react", "sourcemap", "transforms", "typescript", "view", "visit"] }
 deno_core = { version = "0.110.0", path = "../core" }
-deno_doc = "=0.22.0"
-deno_graph = "=0.13.0"
-deno_lint = { version = "=0.20.0", features = ["docs"] }
+deno_doc = "0.22.0"
+deno_graph = "0.13.0"
+deno_lint = { version = "0.20.0", features = ["docs"] }
 deno_runtime = { version = "0.36.0", path = "../runtime" }
 
 atty = "=0.2.14"

--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -39,26 +39,26 @@ winapi = "=0.3.9"
 winres = "=0.1.11"
 
 [dependencies]
-deno_ast = { version = "0.5.0", features = ["bundler", "codegen", "dep_graph", "module_specifier", "proposal", "react", "sourcemap", "transforms", "typescript", "view", "visit"] }
+deno_ast = { version = "=0.7.0", features = ["bundler", "codegen", "dep_graph", "module_specifier", "proposal", "react", "sourcemap", "transforms", "typescript", "view", "visit"] }
 deno_core = { version = "0.110.0", path = "../core" }
-deno_doc = "0.21.0"
-deno_graph = "0.12.0"
-deno_lint = { version = "0.19.0", features = ["docs"] }
+deno_doc = "=0.22.0"
+deno_graph = "=0.13.0"
+deno_lint = { version = "=0.20.0", features = ["docs"] }
 deno_runtime = { version = "0.36.0", path = "../runtime" }
 
 atty = "=0.2.14"
 base64 = "=0.13.0"
 clap = "=2.33.3"
-data-url = "=0.1.0"
+data-url = "=0.1.1"
 dissimilar = "=1.0.2"
 dprint-plugin-json = "=0.13.2"
 dprint-plugin-markdown = "=0.11.3"
-dprint-plugin-typescript = "=0.59.2"
+dprint-plugin-typescript = "=0.60.0"
 encoding_rs = "=0.8.29"
 env_logger = "=0.8.4"
 fancy-regex = "=0.7.1"
 http = "=0.2.4"
-import_map = "=0.3.3"
+import_map = "=0.4.0"
 jsonc-parser = { version = "=0.17.0", features = ["serde"] }
 lazy_static = "=1.4.0"
 libc = "=0.2.106"

--- a/cli/ast/mod.rs
+++ b/cli/ast/mod.rs
@@ -25,6 +25,7 @@ use deno_ast::swc::transforms::hygiene;
 use deno_ast::swc::transforms::pass::Optional;
 use deno_ast::swc::transforms::proposals;
 use deno_ast::swc::transforms::react;
+use deno_ast::swc::transforms::resolver::ts_resolver;
 use deno_ast::swc::transforms::resolver_with_mark;
 use deno_ast::swc::transforms::typescript;
 use deno_ast::swc::visit::FoldWith;
@@ -333,7 +334,7 @@ pub fn transpile_module(
   let input = StringInput::from(&*source_file);
   let comments = SingleThreadedComments::default();
   let syntax = get_syntax(media_type);
-  let lexer = Lexer::new(syntax, deno_ast::TARGET, input, Some(&comments));
+  let lexer = Lexer::new(syntax, deno_ast::ES_VERSION, input, Some(&comments));
   let mut parser = deno_ast::swc::parser::Parser::new_from(lexer);
   let module = parser.parse_module().map_err(|err| {
     let location = cm.lookup_char_pos(err.span().lo);
@@ -420,6 +421,7 @@ fn fold_program(
     ),
   );
   let mut passes = chain!(
+    ts_resolver(top_level_mark),
     Optional::new(transforms::DownlevelImportsFolder, options.repl_imports),
     Optional::new(transforms::StripExportsFolder, options.repl_imports),
     proposals::decorators::decorators(proposals::decorators::Config {
@@ -512,34 +514,64 @@ mod tests {
     let specifier = resolve_url_or_path("https://deno.land/x/mod.ts")
       .expect("could not resolve specifier");
     let source = r#"
-    enum D {
-      A,
-      B,
-      C,
-    }
+enum D {
+  A,
+  B,
+}
 
-    export class A {
-      private b: string;
-      protected c: number = 1;
-      e: "foo";
-      constructor (public d = D.A) {
-        const e = "foo" as const;
-        this.e = e;
-      }
-    }
+namespace Test {
+  export const Value = 5;
+}
+
+class A {
+  private b: string;
+  protected c: number = 1;
+  e: "foo";
+  constructor (public d = D.A) {
+    const e = "foo" as const;
+    this.e = e;
+    console.log(Test.Value);
+  }
+
+  method() {
+  }
+}
     "#;
-    let module = parse_module(ParseParams {
+    let module = deno_ast::parse_script(ParseParams {
       specifier: specifier.as_str().to_string(),
       source: SourceTextInfo::from_string(source.to_string()),
       media_type: deno_ast::MediaType::TypeScript,
       capture_tokens: false,
       maybe_syntax: None,
       scope_analysis: false,
-    })
-    .expect("could not parse module");
+    }).unwrap();
     let (code, maybe_map) = transpile(&module, &EmitOptions::default())
       .expect("could not strip types");
-    assert!(code.starts_with("var D;\n(function(D) {\n"));
+    let expected_text = r#"var D;
+(function(D) {
+    D[D["A"] = 0] = "A";
+    D[D["B"] = 1] = "B";
+})(D || (D = {
+}));
+var Test;
+(function(Test) {
+    Test.Value = 5;
+})(Test || (Test = {
+}));
+export class A {
+    d;
+    b;
+    c = 1;
+    e;
+    constructor(d = D.A){
+        this.d = d;
+        const e = "foo";
+        this.e = e;
+        console.log(Test.Value);
+    }
+}
+"#;
+    assert_eq!(&code[..expected_text.len()], expected_text);
     assert!(
       code.contains("\n//# sourceMappingURL=data:application/json;base64,")
     );

--- a/cli/ast/mod.rs
+++ b/cli/ast/mod.rs
@@ -17,6 +17,8 @@ use deno_ast::swc::common::Globals;
 use deno_ast::swc::common::Mark;
 use deno_ast::swc::common::SourceMap;
 use deno_ast::swc::common::Spanned;
+use deno_ast::swc::parser::error::Error as SwcError;
+use deno_ast::swc::parser::error::SyntaxError;
 use deno_ast::swc::parser::lexer::Lexer;
 use deno_ast::swc::parser::StringInput;
 use deno_ast::swc::transforms::fixer;
@@ -38,6 +40,7 @@ use deno_core::resolve_url_or_path;
 use deno_core::serde_json;
 use deno_core::ModuleSpecifier;
 use std::cell::RefCell;
+use std::fmt;
 use std::rc::Rc;
 
 mod bundle_hook;
@@ -100,6 +103,25 @@ impl From<Location> for ModuleSpecifier {
 impl std::fmt::Display for Location {
   fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
     write!(f, "{}:{}:{}", self.specifier, self.line, self.col)
+  }
+}
+
+#[derive(Debug)]
+pub struct Diagnostics(pub Vec<Diagnostic>);
+
+impl std::error::Error for Diagnostics {}
+
+impl fmt::Display for Diagnostics {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    for (i, diagnostic) in self.0.iter().enumerate() {
+      if i > 0 {
+        write!(f, "\n\n")?;
+      }
+
+      write!(f, "{}", diagnostic)?
+    }
+
+    Ok(())
   }
 }
 
@@ -259,6 +281,7 @@ pub fn transpile(
   parsed_source: &ParsedSource,
   options: &EmitOptions,
 ) -> Result<(String, Option<String>), AnyError> {
+  ensure_no_fatal_diagnostics(parsed_source.diagnostics().iter())?;
   let program: Program = (*parsed_source.program()).clone();
   let source_map = Rc::new(SourceMap::default());
   let source_map_config = SourceMapConfig {
@@ -335,18 +358,16 @@ pub fn transpile_module(
   let syntax = get_syntax(media_type);
   let lexer = Lexer::new(syntax, deno_ast::ES_VERSION, input, Some(&comments));
   let mut parser = deno_ast::swc::parser::Parser::new_from(lexer);
-  let module = parser.parse_module().map_err(|err| {
-    let location = cm.lookup_char_pos(err.span().lo);
-    Diagnostic {
-      specifier: specifier.to_string(),
-      span: err.span(),
-      display_position: LineAndColumnDisplay {
-        line_number: location.line,
-        column_number: location.col_display + 1,
-      },
-      kind: err.into_kind(),
-    }
-  })?;
+  let module = parser
+    .parse_module()
+    .map_err(|e| swc_err_to_diagnostic(&cm, specifier, e))?;
+  let diagnostics = parser
+    .take_errors()
+    .into_iter()
+    .map(|e| swc_err_to_diagnostic(&cm, specifier, e))
+    .collect::<Vec<_>>();
+
+  ensure_no_fatal_diagnostics(diagnostics.iter())?;
 
   let top_level_mark = Mark::fresh(Mark::root());
   let program = fold_program(
@@ -426,9 +447,10 @@ fn fold_program(
     helpers::inject_helpers(),
     resolver_with_mark(top_level_mark),
     Optional::new(
-      typescript::strip::strip_with_config(strip_config_from_emit_options(
-        options
-      ), top_level_mark),
+      typescript::strip::strip_with_config(
+        strip_config_from_emit_options(options),
+        top_level_mark
+      ),
       !options.transform_jsx
     ),
     Optional::new(
@@ -455,18 +477,32 @@ fn fold_program(
   });
 
   let diagnostics = diagnostics_cell.borrow();
-  if let Some(diagnostic) = diagnostics.iter().find(|d| is_fatal_diagnostic(d))
-  {
+  ensure_no_fatal_swc_diagnostics(&source_map, diagnostics.iter())?;
+  Ok(result)
+}
+
+fn ensure_no_fatal_swc_diagnostics<'a>(
+  source_map: &SourceMap,
+  diagnostics: impl Iterator<Item = &'a SwcDiagnostic>,
+) -> Result<(), AnyError> {
+  let fatal_diagnostics = diagnostics
+    .filter(|d| is_fatal_swc_diagnostic(d))
+    .collect::<Vec<_>>();
+  if !fatal_diagnostics.is_empty() {
     Err(anyhow!(
       "{}",
-      format_swc_diagnostic(&source_map, diagnostic)
+      fatal_diagnostics
+        .iter()
+        .map(|d| format_swc_diagnostic(source_map, d))
+        .collect::<Vec<_>>()
+        .join("\n\n")
     ))
   } else {
-    Ok(result)
+    Ok(())
   }
 }
 
-fn is_fatal_diagnostic(diagnostic: &SwcDiagnostic) -> bool {
+fn is_fatal_swc_diagnostic(diagnostic: &SwcDiagnostic) -> bool {
   use deno_ast::swc::common::errors::Level;
   match diagnostic.level {
     Level::Bug
@@ -496,6 +532,51 @@ fn format_swc_diagnostic(
   } else {
     diagnostic.message()
   }
+}
+
+fn swc_err_to_diagnostic(
+  source_map: &SourceMap,
+  specifier: &ModuleSpecifier,
+  err: SwcError,
+) -> Diagnostic {
+  let location = source_map.lookup_char_pos(err.span().lo);
+  Diagnostic {
+    specifier: specifier.to_string(),
+    span: err.span(),
+    display_position: LineAndColumnDisplay {
+      line_number: location.line,
+      column_number: location.col_display + 1,
+    },
+    kind: err.into_kind(),
+  }
+}
+
+fn ensure_no_fatal_diagnostics<'a>(
+  diagnostics: impl Iterator<Item = &'a Diagnostic>,
+) -> Result<(), Diagnostics> {
+  let fatal_diagnostics = diagnostics
+    .filter(|d| is_fatal_syntax_error(&d.kind))
+    .map(ToOwned::to_owned)
+    .collect::<Vec<_>>();
+  if !fatal_diagnostics.is_empty() {
+    Err(Diagnostics(fatal_diagnostics))
+  } else {
+    Ok(())
+  }
+}
+
+fn is_fatal_syntax_error(error_kind: &SyntaxError) -> bool {
+  matches!(
+    error_kind,
+    // expected identifier
+    SyntaxError::TS1003 |
+        // expected semi-colon
+        SyntaxError::TS1005 |
+        // expected expression
+        SyntaxError::TS1109 |
+        // unterminated string literal
+        SyntaxError::UnterminatedStrLit
+  )
 }
 
 #[cfg(test)]
@@ -542,7 +623,8 @@ export class A {
       capture_tokens: false,
       maybe_syntax: None,
       scope_analysis: false,
-    }).unwrap();
+    })
+    .unwrap();
     let (code, maybe_map) = transpile(&module, &EmitOptions::default())
       .expect("could not strip types");
     let expected_text = r#"var D;

--- a/cli/ast/mod.rs
+++ b/cli/ast/mod.rs
@@ -25,7 +25,6 @@ use deno_ast::swc::transforms::hygiene;
 use deno_ast::swc::transforms::pass::Optional;
 use deno_ast::swc::transforms::proposals;
 use deno_ast::swc::transforms::react;
-use deno_ast::swc::transforms::resolver::ts_resolver;
 use deno_ast::swc::transforms::resolver_with_mark;
 use deno_ast::swc::transforms::typescript;
 use deno_ast::swc::visit::FoldWith;
@@ -397,31 +396,27 @@ fn fold_program(
   comments: &SingleThreadedComments,
   top_level_mark: Mark,
 ) -> Result<Program, AnyError> {
-  let jsx_pass = chain!(
-    resolver_with_mark(top_level_mark),
-    react::react(
-      source_map.clone(),
-      Some(comments),
-      react::Options {
-        pragma: options.jsx_factory.clone(),
-        pragma_frag: options.jsx_fragment_factory.clone(),
-        // this will use `Object.assign()` instead of the `_extends` helper
-        // when spreading props.
-        use_builtins: true,
-        runtime: if options.jsx_automatic {
-          Some(react::Runtime::Automatic)
-        } else {
-          None
-        },
-        development: options.jsx_development,
-        import_source: options.jsx_import_source.clone().unwrap_or_default(),
-        ..Default::default()
+  let jsx_pass = react::react(
+    source_map.clone(),
+    Some(comments),
+    react::Options {
+      pragma: options.jsx_factory.clone(),
+      pragma_frag: options.jsx_fragment_factory.clone(),
+      // this will use `Object.assign()` instead of the `_extends` helper
+      // when spreading props.
+      use_builtins: true,
+      runtime: if options.jsx_automatic {
+        Some(react::Runtime::Automatic)
+      } else {
+        None
       },
-      top_level_mark,
-    ),
+      development: options.jsx_development,
+      import_source: options.jsx_import_source.clone().unwrap_or_default(),
+      ..Default::default()
+    },
+    top_level_mark,
   );
   let mut passes = chain!(
-    ts_resolver(top_level_mark),
     Optional::new(transforms::DownlevelImportsFolder, options.repl_imports),
     Optional::new(transforms::StripExportsFolder, options.repl_imports),
     proposals::decorators::decorators(proposals::decorators::Config {
@@ -429,10 +424,11 @@ fn fold_program(
       emit_metadata: options.emit_metadata
     }),
     helpers::inject_helpers(),
+    resolver_with_mark(top_level_mark),
     Optional::new(
       typescript::strip::strip_with_config(strip_config_from_emit_options(
         options
-      )),
+      ), top_level_mark),
       !options.transform_jsx
     ),
     Optional::new(
@@ -509,6 +505,8 @@ mod tests {
   use deno_ast::ParseParams;
   use deno_ast::SourceTextInfo;
 
+  use pretty_assertions::assert_eq;
+
   #[test]
   fn test_transpile() {
     let specifier = resolve_url_or_path("https://deno.land/x/mod.ts")
@@ -519,25 +517,25 @@ enum D {
   B,
 }
 
-namespace Test {
+namespace N {
+  export enum D {
+    A = "value"
+  }
   export const Value = 5;
 }
 
-class A {
+export class A {
   private b: string;
   protected c: number = 1;
   e: "foo";
   constructor (public d = D.A) {
     const e = "foo" as const;
     this.e = e;
-    console.log(Test.Value);
-  }
-
-  method() {
+    console.log(N.Value);
   }
 }
     "#;
-    let module = deno_ast::parse_script(ParseParams {
+    let module = deno_ast::parse_module(ParseParams {
       specifier: specifier.as_str().to_string(),
       source: SourceTextInfo::from_string(source.to_string()),
       media_type: deno_ast::MediaType::TypeScript,
@@ -553,10 +551,15 @@ class A {
     D[D["B"] = 1] = "B";
 })(D || (D = {
 }));
-var Test;
-(function(Test) {
-    Test.Value = 5;
-})(Test || (Test = {
+var N;
+(function(N1) {
+    let D;
+    (function(D) {
+        D["A"] = "value";
+    })(D = N1.D || (N1.D = {
+    }));
+    N1.Value = 5;
+})(N || (N = {
 }));
 export class A {
     d;
@@ -567,7 +570,7 @@ export class A {
         this.d = d;
         const e = "foo";
         this.e = e;
-        console.log(Test.Value);
+        console.log(N.Value);
     }
 }
 "#;

--- a/cli/http_util.rs
+++ b/cli/http_util.rs
@@ -149,7 +149,6 @@ mod tests {
   use super::*;
   use crate::version;
   use deno_runtime::deno_fetch::create_http_client;
-  use deno_runtime::deno_tls::rustls::RootCertStore;
   use std::fs::read;
 
   fn create_test_client() -> Client {
@@ -409,6 +408,8 @@ mod tests {
   #[cfg(not(windows))]
   #[tokio::test]
   async fn test_fetch_with_empty_certificate_store() {
+    use deno_runtime::deno_tls::rustls::RootCertStore;
+
     let _http_server_guard = test_util::http_server();
     // Relies on external http server with a valid mozilla root CA cert.
     let url = Url::parse("https://deno.land").unwrap();

--- a/cli/lsp/code_lens.rs
+++ b/cli/lsp/code_lens.rs
@@ -9,7 +9,6 @@ use super::tsc::NavigationTree;
 
 use deno_ast::swc::ast;
 use deno_ast::swc::common::Span;
-use deno_ast::swc::visit::Node;
 use deno_ast::swc::visit::Visit;
 use deno_ast::swc::visit::VisitWith;
 use deno_ast::ParsedSource;
@@ -129,7 +128,7 @@ impl DenoTestCollector {
 }
 
 impl Visit for DenoTestCollector {
-  fn visit_call_expr(&mut self, node: &ast::CallExpr, _parent: &dyn Node) {
+  fn visit_call_expr(&mut self, node: &ast::CallExpr) {
     if let ast::ExprOrSuper::Expr(callee_expr) = &node.callee {
       match callee_expr.as_ref() {
         ast::Expr::Ident(ident) => {
@@ -155,7 +154,7 @@ impl Visit for DenoTestCollector {
     }
   }
 
-  fn visit_var_decl(&mut self, node: &ast::VarDecl, _parent: &dyn Node) {
+  fn visit_var_decl(&mut self, node: &ast::VarDecl) {
     for decl in &node.decls {
       if let Some(init) = &decl.init {
         match init.as_ref() {
@@ -401,12 +400,7 @@ fn collect_test(
     if let Some(parsed_source) = parsed_source {
       let mut collector =
         DenoTestCollector::new(specifier.clone(), parsed_source.clone());
-      parsed_source.module().visit_with(
-        &ast::Invalid {
-          span: deno_ast::swc::common::DUMMY_SP,
-        },
-        &mut collector,
-      );
+      parsed_source.module().visit_with(&mut collector);
       return Ok(collector.take());
     }
   }
@@ -564,12 +558,7 @@ mod tests {
     .unwrap();
     let mut collector =
       DenoTestCollector::new(specifier, parsed_module.clone());
-    parsed_module.module().visit_with(
-      &ast::Invalid {
-        span: deno_ast::swc::common::DUMMY_SP,
-      },
-      &mut collector,
-    );
+    parsed_module.module().visit_with(&mut collector);
     assert_eq!(
       collector.take(),
       vec![

--- a/cli/tests/integration/repl_tests.rs
+++ b/cli/tests/integration/repl_tests.rs
@@ -267,7 +267,8 @@ fn typescript_declarations() {
     Some(vec![("NO_COLOR".to_owned(), "1".to_owned())]),
     false,
   );
-  assert!(out.ends_with("undefined\n0\n2\nundefined\nundefined\n"));
+  let expected_end_text = "undefined\n0\n2\nundefined\nundefined\n";
+  assert_eq!(&out[out.len() - expected_end_text.len()..], expected_end_text);
   assert!(err.is_empty());
 }
 

--- a/cli/tests/integration/repl_tests.rs
+++ b/cli/tests/integration/repl_tests.rs
@@ -63,12 +63,14 @@ fn pty_bad_input() {
 fn pty_syntax_error_input() {
   util::with_pty(&["repl"], |mut console| {
     console.write_line("('\\u')");
-    console.write_line("('");
+    console.write_line("'");
+    console.write_line("[{'a'}];");
     console.write_line("close();");
 
     let output = console.read_all_output();
+    assert!(output.contains("Expected 4 hex characters"));
     assert!(output.contains("Unterminated string constant"));
-    assert!(output.contains("Unexpected eof"));
+    assert!(output.contains("Expected a semicolon"));
   });
 }
 
@@ -268,7 +270,10 @@ fn typescript_declarations() {
     false,
   );
   let expected_end_text = "undefined\n0\n2\nundefined\nundefined\n";
-  assert_eq!(&out[out.len() - expected_end_text.len()..], expected_end_text);
+  assert_eq!(
+    &out[out.len() - expected_end_text.len()..],
+    expected_end_text
+  );
   assert!(err.is_empty());
 }
 

--- a/cli/tests/testdata/bundle.test.out
+++ b/cli/tests/testdata/bundle.test.out
@@ -8,20 +8,20 @@ function returnsFoo() {
 function printHello2() {
     printHello();
 }
-function returnsHi1() {
+function returnsHi() {
     return "Hi";
 }
-function returnsFoo21() {
+function returnsFoo2() {
     return returnsFoo();
 }
-function printHello31() {
+function printHello3() {
     printHello2();
 }
-function throwsError1() {
+function throwsError() {
     throw Error("exception from mod1");
 }
-export { returnsHi1 as returnsHi };
-export { returnsFoo21 as returnsFoo2 };
-export { printHello31 as printHello3 };
-export { throwsError1 as throwsError };
+export { returnsHi as returnsHi };
+export { returnsFoo2 as returnsFoo2 };
+export { printHello3 as printHello3 };
+export { throwsError as throwsError };
 

--- a/cli/tests/testdata/compiler_api_test.ts
+++ b/cli/tests/testdata/compiler_api_test.ts
@@ -2,8 +2,8 @@
 import {
   assert,
   assertEquals,
+  assertRejects,
   assertStringIncludes,
-  assertThrowsAsync,
 } from "../../../test_util/std/testing/asserts.ts";
 
 Deno.test({
@@ -268,7 +268,7 @@ Deno.test({
       Object.keys(files).sort(),
       ["deno:///bundle.js", "deno:///bundle.js.map"].sort(),
     );
-    assert(files["deno:///bundle.js"].includes(`const bar1 = "bar"`));
+    assert(files["deno:///bundle.js"].includes(`const bar = "bar"`));
   },
 });
 
@@ -312,7 +312,7 @@ Deno.test({
       Object.keys(files).sort(),
       ["deno:///bundle.js.map", "deno:///bundle.js"].sort(),
     );
-    assert(files["deno:///bundle.js"].includes(`const bar1 = "bar"`));
+    assert(files["deno:///bundle.js"].includes(`const bar = "bar"`));
   },
 });
 
@@ -430,7 +430,7 @@ Deno.test({
 Deno.test({
   name: `Deno.emit() - throws descriptive error when unable to load import map`,
   async fn() {
-    await assertThrowsAsync(
+    await assertRejects(
       async () => {
         await Deno.emit("/a.ts", {
           bundle: "classic",
@@ -566,7 +566,7 @@ Deno.test({
       {
         sources: {
           "file:///a.tsx": `/** @jsxImportSource https://example.com/jsx */
-          
+
           export function App() {
             return (
               <div><></></div>

--- a/cli/tools/repl/mod.rs
+++ b/cli/tools/repl/mod.rs
@@ -1,6 +1,7 @@
 // Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
 
 use crate::ast::transpile;
+use crate::ast::Diagnostics;
 use crate::ast::ImportsNotUsedAsValues;
 use crate::colors;
 use crate::proc_state::ProcState;
@@ -507,6 +508,16 @@ impl ReplSession {
     &mut self,
     line: &str,
   ) -> Result<EvaluationOutput, AnyError> {
+    fn format_diagnostic(diagnostic: &deno_ast::Diagnostic) -> String {
+      format!(
+        "{}: {} at {}:{}",
+        colors::red("parse error"),
+        diagnostic.message(),
+        diagnostic.display_position.line_number,
+        diagnostic.display_position.column_number,
+      )
+    }
+
     match self.evaluate_line_with_object_wrapping(line).await {
       Ok(evaluate_response) => {
         let evaluate_result = evaluate_response.get("result").unwrap();
@@ -528,14 +539,20 @@ impl ReplSession {
       Err(err) => {
         // handle a parsing diagnostic
         match err.downcast_ref::<deno_ast::Diagnostic>() {
-          Some(diagnostic) => Ok(EvaluationOutput::Error(format!(
-            "{}: {} at {}:{}",
-            colors::red("parse error"),
-            diagnostic.message(),
-            diagnostic.display_position.line_number,
-            diagnostic.display_position.column_number,
-          ))),
-          None => Err(err),
+          Some(diagnostic) => {
+            Ok(EvaluationOutput::Error(format_diagnostic(diagnostic)))
+          }
+          None => match err.downcast_ref::<Diagnostics>() {
+            Some(diagnostics) => Ok(EvaluationOutput::Error(
+              diagnostics
+                .0
+                .iter()
+                .map(format_diagnostic)
+                .collect::<Vec<_>>()
+                .join("\n\n"),
+            )),
+            None => Err(err),
+          },
         }
       }
     }
@@ -545,8 +562,8 @@ impl ReplSession {
     &mut self,
     line: &str,
   ) -> Result<Value, AnyError> {
-    // It is a bit unexpected that { "foo": "bar" } is interpreted as a block
-    // statement rather than an object literal so we interpret it as an expression statement
+    // Expressions like { "foo": "bar" } are interpreted as block expressions at the
+    // statement level rather than an object literal so we interpret it as an expression statement
     // to match the behavior found in a typical prompt including browser developer tools.
     let wrapped_line = if line.trim_start().starts_with('{')
       && !line.trim_end().ends_with(';')
@@ -556,20 +573,22 @@ impl ReplSession {
       line.to_string()
     };
 
-    let evaluate_response = self.evaluate_ts_expression(&wrapped_line).await?;
+    let evaluate_response = self.evaluate_ts_expression(&wrapped_line).await;
 
     // If that fails, we retry it without wrapping in parens letting the error bubble up to the
     // user if it is still an error.
-    let evaluate_response =
-      if evaluate_response.get("exceptionDetails").is_some()
-        && wrapped_line != line
-      {
-        self.evaluate_ts_expression(line).await?
-      } else {
-        evaluate_response
-      };
-
-    Ok(evaluate_response)
+    if wrapped_line != line
+      && (evaluate_response.is_err()
+        || evaluate_response
+          .as_ref()
+          .unwrap()
+          .get("exceptionDetails")
+          .is_some())
+    {
+      self.evaluate_ts_expression(line).await
+    } else {
+      evaluate_response
+    }
   }
 
   async fn set_last_thrown_error(

--- a/core/Cargo.toml
+++ b/core/Cargo.toml
@@ -15,8 +15,7 @@ path = "lib.rs"
 [dependencies]
 anyhow = "1.0.43"
 futures = "0.3.16"
-# TODO(lucacasonato): unlock when https://github.com/tkaitchuck/aHash/issues/95 is resolved
-indexmap = "=1.6.2"
+indexmap = "1.7.0"
 lazy_static = "1.4.0"
 libc = "0.2.106"
 log = "0.4.14"

